Modified kcp version to release-0.5

xinredhat opened the pull request:
**Changes**
1. Modified KCP image to [release-0.5](https://github.com/kcp-dev/kcp/pkgs/container/kcp/24100239?tag=release-0.5)
2. Modified  KCP Syncer image to [release-0.5](https://github.com/kcp-dev/kcp/pkgs/container/kcp%2Fsyncer/24099863?tag=release-0.5 )
3. Modified ./ckcp/test.sh to look for namespace with proper label
4. Modified ./ckcp/test.sh for PipelineRun case by removing serviceAccountName property

**Known Issue**

1. When executing `./ckcp/test.sh pipelines`, the PipelineRun, such like `test-pipelinerun-xxxx` will fail due to the SA token secret being renamed from `default-token-xxxx` in icp to  `kcp-default-token-xxx` on the workload cluster.  It couldn't find the secret.
2. `./ckcp/test.sh` couldn't run the trigger cases, It needs to expose the event listeners through a global load balancer fronting kcp to handle  redirection of the requests


**Testing Locally**
1. Execute `./ckcp/test.sh pipelines` to runs TaskRun and PipelineRun cases

fgiloux commented:
Hi @xinredhat. Thanks for looking at that.
There are a few more places that would need to be changed:
https://github.com/openshift-pipelines/pipelines-service/blob/main/images/kcp-registrar/register.sh#L127
https://github.com/openshift-pipelines/pipelines-service/blob/main/images/kcp-registrar/Dockerfile#L8
https://github.com/openshift-pipelines/pipelines-service/blob/main/local/kcp/start.sh#L43

Also could you please modify test.sh so that it deploy TaskRuns PipelineRuns that either don't have a serviceAccountName set or use pipeline SA.

As part of this PR I would like this to get documented, possibly in a new document "phase 1 limitations". There we should stipulate:
- not specifying an SA is OK
- pipeline SA can be used
- default SA does not work and would point to kcp API
- custom SA created in kcp workspace may have the same flaw for our phase 1 approach
Describing the behavior and the challenge:
- The default SA in kcp workspace has the secret containing its token synchronised in the workload cluster but under a different name (with kcp prefix). In phase 1 we don't want Tekton pods to communicate with the kcp API. As such they should not use an SA and its token created by kcp. This is a limitation but existing SA from the workload cluster can be used instead.

xinredhat commented:
@fgiloux sure will do. thanks for the clarification.

Roming22 commented on ckcp/openshift_dev_setup.sh:
```diff
@@ -227,7 +227,12 @@ patches:
+    echo -n "."
+    sleep 5
+  done
+  kubectl exec -n $APP $podname -- cat /etc/kcp/config/admin.kubeconfig > "$KUBECONFIG_KCP" 
```
Move the `echo "OK"` line 220 after this line.
Also I think that there's a `kubectl cp` command.

xinredhat replied:
>Move the echo "OK" line 220 after this line.

Fixed

bnallapeta replied:
@xinredhat I did not have an issue using kubectl cp with the same path /etc/kcp/config/admin.kubeconfig. Not sure where the /data folder is coming into picture for copying the kubeconfig.

```
++ kubectl exec -n ckcp ckcp-668bf6b867-xc28q -- ls /etc/kcp/config/admin.kubeconfig
++ echo 0
+ [[ 0 -ne 0 ]]
+ kubectl cp ckcp/ckcp-668bf6b867-xc28q:/etc/kcp/config/admin.kubeconfig /home/bnr/workspace/pipelines-service/ckcp/work/kubeconfig/admin.kubeconfig
+ echo OK

```

xinredhat replied:
`kubectl cp` modification is done

xinredhat commented:
> Hi @xinredhat. Thanks for looking at that. There are a few more places that would need to be changed: https://github.com/openshift-pipelines/pipelines-service/blob/main/images/kcp-registrar/register.sh#L127 https://github.com/openshift-pipelines/pipelines-service/blob/main/images/kcp-registrar/Dockerfile#L8 https://github.com/openshift-pipelines/pipelines-service/blob/main/local/kcp/start.sh#L43

Fixed

> Also could you please modify test.sh so that it deploy TaskRuns PipelineRuns that either don't have a serviceAccountName set or use pipeline SA.

Fixed by removing serviceAccountName set

xinredhat commented on ckcp/openshift_dev_setup.sh:
```diff
@@ -15,7 +15,7 @@ WORK_DIR="$SCRIPT_DIR/work"
 KUBECONFIG_KCP="$WORK_DIR/kubeconfig/admin.kubeconfig"
 KUBECONFIG_MERGED="merged-config.kubeconfig:$KUBECONFIG:$KUBECONFIG_KCP"
-
+KCP_VERSION="release-0.5"
```
fixed

bnallapeta replied:
yes, something like this should work. With this change, in the future, if we need to upgrade to release-0.6 or anything else, we need to change it only at pipelines-service/ckcp/openshift/overlays/dev/kustomization.yaml and nowhere else. So, what we can do is, we can extract this tag from this file and use it everywhere else.

`kcp-image-tag=yq '.images[].newTag' < pipelines-service/ckcp/openshift/overlays/dev/kustomization.yaml`

bnallapeta commented on ckcp/openshift_dev_setup.sh:
```diff
@@ -254,8 +258,9 @@ patches:
       KUBECONFIG="$KUBECONFIG_KCP" kubectl kcp workload sync local \
-      --syncer-image ghcr.io/kcp-dev/kcp/syncer:release-0.4 \
+      --syncer-image ghcr.io/kcp-dev/kcp/syncer:$KCP_VERSION \
       --resources conditions.tekton.dev,pipelines.tekton.dev,pipelineruns.tekton.dev,pipelineresources.tekton.dev,runs.tekton.dev,tasks.tekton.dev,taskruns.tekton.dev > "$kube_dir/syncer.yaml"
```
As discussed, we must also add tekton triggers resources here to sync without which ./test.sh fails for triggers examples.

bnallapeta replied:
`+ echo 'Arg triggers passed. Running triggers tests...'
Arg triggers passed. Running triggers tests...
+ echo 'Simulating a Github PR through a curl request which creates a TaskRun (from tektoncd/triggers/examples)'
Simulating a Github PR through a curl request which creates a TaskRun (from tektoncd/triggers/examples)
+ KUBECONFIG=/home/bnr/workspace/pipelines-service/ckcp/work/kubeconfig/admin.kubeconfig
+ kubectl apply -f https://raw.githubusercontent.com/tektoncd/triggers/v0.18.0/examples/v1beta1/github/github-eventlistener-interceptor.yaml
unable to recognize "https://raw.githubusercontent.com/tektoncd/triggers/v0.18.0/examples/v1beta1/github/github-eventlistener-interceptor.yaml": no matches for kind "EventListener" in version "triggers.tekton.dev/v1beta1"
unable to recognize "https://raw.githubusercontent.com/tektoncd/triggers/v0.18.0/examples/v1beta1/github/github-eventlistener-interceptor.yaml": no matches for kind "TriggerBinding" in version "triggers.tekton.dev/v1beta1"
unable to recognize "https://raw.githubusercontent.com/tektoncd/triggers/v0.18.0/examples/v1beta1/github/github-eventlistener-interceptor.yaml": no matches for kind "TriggerTemplate" in version "triggers.tekton.dev/v1beta1"
`

xinredhat replied:
As discussed that we remove the trigger resources to the way it was before (just the pipeline resources)

bnallapeta reviewed (commented):
https://github.com/openshift-pipelines/pipelines-service/blob/main/ckcp/openshift_dev_setup.sh - cert-manager plugin

Could we add some meaningful error message either [here](https://github.com/openshift-pipelines/pipelines-service/blob/main/ckcp/openshift_dev_setup.sh#L160) or as a pre-requisite check before starting the rest of the script to check that all the pre-requisites are satisfied.

@ramessesii2 has created #104 for the same.

Roming22 commented on ckcp/openshift_dev_setup.sh:
```diff
@@ -240,9 +245,8 @@ patches:
   echo -n "  - Workspace: "
   # It's not allowed to create WorkloadCluster resource in a non-universal workspace
   if ! KUBECONFIG="$KUBECONFIG_KCP" kubectl get workspaces demo >/dev/null 2>&1; then
```
Because there's no else, I believe there will a bug if the `demo` workspace exists, but we're in another workspace at the moment.
1. Make sure we're in the root workspace.
2. Move the `kubectl kcp workspaces use demo` out of the `if`.

xinredhat replied:
fixed.

Roming22 commented on images/kcp-registrar/register.sh:
```diff
@@ -124,7 +124,7 @@ register() {
             printf "Registering cluster\n"
             KUBECONFIG=${kcp_kcfg} kubectl kcp workload sync "${clusters[$i]}" \
-                 --syncer-image ghcr.io/kcp-dev/kcp/syncer:release-0.4 \
+                 --syncer-image ghcr.io/kcp-dev/kcp/syncer:KCP_BRANCH \
```
Could we use an env variable instead? The container should have the variable set when the image is built, and that it would allow to run the script locally without having to modify it.

xinredhat replied:
Good idea, will try

xinredhat commented:
> https://github.com/openshift-pipelines/pipelines-service/blob/main/ckcp/openshift_dev_setup.sh - cert-manager plugin
> 
> Could we add some meaningful error message either [here](https://github.com/openshift-pipelines/pipelines-service/blob/main/ckcp/openshift_dev_setup.sh#L160) or as a pre-requisite check before starting the rest of the script to check that all the pre-requisites are satisfied.
> 
> @ramessesii2 has created #104 for the same.

Fixed

fgiloux commented on docs/phase1_limitations.md:
```diff
@@ -0,0 +1,11 @@
+* Not specifying an SA is OK
+* pipeline SA can be used
+* default SA does not work and would point to kcp API
+* Custom SA created in kcp workspace may have the same flaw for our phase 1 approach
```
=> A custom SA created in kcp workspace will present the same flaw with phase 1.

xinredhat replied:
Fixed

fgiloux commented on local/kcp/start.sh:
```diff
@@ -32,15 +32,17 @@ precheck() {
+kcp-binaries() {
   if [[ -z "${KCP_DIR}" ]]; then
+    CKCP_DIR="${PARENT_PATH}/../../ckcp"
+    kcp_image_tag="$(yq -e '.images[0].newTag' "$CKCP_DIR"/openshift/overlays/dev/kustomization.yaml)" 
```
We don't need that here and having in the script a dependency to ckcp and an openshift overlay does not seem right.
I would just let it as it was. At the end of the day "release-0.5" is just a default value for the kcp branch to use and the developer is free to select a different one.

xinredhat replied:
Fixed.

fgiloux reviewed (changes requested):
Hi @xinredhat thank you for moving this forward. And sorry I feel that the scope of this PR has crept. I still have a few comments. Feel free to ping me if you want to discuss them.

xinredhat commented on docs/phase1_limitations.md:
```diff
@@ -0,0 +1,11 @@
+## Cases Instructions
+* Not specifying an SA is OK
+* pipeline SA can be used
+* default SA does not work and would point to kcp API
```
Fixed

xinredhat commented on docs/phase1_limitations.md:
```diff
@@ -0,0 +1,11 @@
+The default SA in kcp workspace has the secret containing its token synchronised in the workload cluster but under a different name (with kcp prefix). In phase 1 we don't want Tekton pods to communicate with the kcp API. As such they should not use an SA and its token created by kcp. This is a limitation but existing SA from the workload cluster can be used instead.
+
+## Cases Instructions
+* Not specifying an SA is OK
```
Fixed

xinredhat commented on ckcp/test.sh:
```diff
@@ -7,6 +7,24 @@ SCRIPT_DIR="$(cd "$(dirname "$0")" >/dev/null ; pwd)"
+  while ! kubectl get ns -o yaml | grep -q "$ns_locator"; do
+    sleep 2
+  done
+  local KCP_NS_ID="$(kubectl get ns -l internal.workload.kcp.dev/cluster=local -o json \
```
Fixed.

xinredhat commented on ckcp/openshift_dev_setup.sh:
```diff
@@ -238,25 +258,26 @@ patches:
     done
-    KUBECONFIG="$KUBECONFIG_KCP" kubectl kcp workspaces use demo
   fi
+  KUBECONFIG="$KUBECONFIG_KCP" kubectl kcp workspaces use demo >/dev/null 2>&1
```
@fgiloux Occasionally I observed that the script execute next step immediately after creating workspace with  the "--enter" option, it failed. I guess workspace needs a little bit more time, so I added logic to check if workspace is ready. To check the newly created workspace, it should go to its parent workspace. that's why the command is not with the "--enter' option

fgiloux replied:
As discussed: could you?
- add a comment before the while: "commands executed directly after create --enter occasionally fail"
- create an issue in the kcp repository. It is a bug in my opinion. I remember discussions around that but could not find any existing issue

xinredhat replied:
sure , agree, this is the bug https://github.com/kcp-dev/kcp/issues/1274

xinredhat commented on ckcp/openshift_dev_setup.sh:
```diff
@@ -85,6 +85,19 @@ parse_args() {
+# To ensure that dependencies are satisfied
+precheck() {
+  if [ $(kubectl plugin list | grep -c "kubectl-kcp") -eq 0 ]  ; then
+    printf "kcp plugin could not found\n"
```
Fixed, thank you for your review :)

xinredhat commented on ckcp/openshift_dev_setup.sh:
```diff
@@ -85,6 +85,19 @@ parse_args() {
+  fi
+
+  if [ $(kubectl plugin list | grep -c "kubectl-cert_manager") -eq 0 ]; then
+    printf "cert_manager plugin could not found\n"
```
Fixed

xinredhat commented on ckcp/openshift_dev_setup.sh:
```diff
@@ -238,25 +258,26 @@ patches:
-      --syncer-image ghcr.io/kcp-dev/kcp/syncer:release-0.4 \
-      --resources conditions.tekton.dev,pipelines.tekton.dev,pipelineruns.tekton.dev,pipelineresources.tekton.dev,runs.tekton.dev,tasks.tekton.dev,taskruns.tekton.dev > "$kube_dir/syncer.yaml"
+      --syncer-image ghcr.io/kcp-dev/kcp/syncer:$kcp_image_tag \
+      --resources conditions.tekton.dev,pipelines.tekton.dev,pipelineruns.tekton.dev,pipelineresources.tekton.dev,runs.tekton.dev,tasks.tekton.dev,taskruns.tekton.dev,repositories.pipelinesascode.tekton.dev,eventlisteners.triggers.tekton.dev,triggerbindings.triggers.tekton.dev,triggers.triggers.tekton.dev,triggertemplates.triggers.tekton.dev > "$kube_dir/syncer.yaml"
```
Fixed by removing the trigger resources

xinredhat commented on docs/phase1_limitations.md:
```diff
@@ -0,0 +1,11 @@
+
+
+## Limitation
+The default SA in kcp workspace has the secret containing its token synchronised in the workload cluster but under a different name (with kcp prefix). In phase 1 we don't want Tekton pods to communicate with the kcp API. As such they should not use an SA and its token created by kcp. This is a limitation but existing SA from the workload cluster can be used instead.
```
Fixed

xinredhat commented on docs/phase1_limitations.md:
```diff
@@ -0,0 +1,11 @@
+# Phase 1 Limitations
+
+
+## Limitation
```
Fixed

xinredhat commented on docs/phase1_limitations.md:
```diff
@@ -0,0 +1,11 @@
+## Limitation
+The default SA in kcp workspace has the secret containing its token synchronised in the workload cluster but under a different name (with kcp prefix). In phase 1 we don't want Tekton pods to communicate with the kcp API. As such they should not use an SA and its token created by kcp. This is a limitation but existing SA from the workload cluster can be used instead.
+
+## Cases Instructions
```
Fixed

xinredhat commented on docs/phase1_limitations.md:
```diff
@@ -0,0 +1,11 @@
+## Limitation
+The default SA in kcp workspace has the secret containing its token synchronised in the workload cluster but under a different name (with kcp prefix). In phase 1 we don't want Tekton pods to communicate with the kcp API. As such they should not use an SA and its token created by kcp. This is a limitation but existing SA from the workload cluster can be used instead.
+
+## Cases Instructions
```
Thank you for your proposal. Fixed

Roming22 commented:
@xinredhat It may be a good idea to jump on a quick call with Bharath and Frederic if there is just a few comments that are blocking the merge of the PR.
